### ISSUE TITLE
[FEATURE] Add support for System.Text.Json serialization

### ISSUE THREAD

**Issue by Scrambles56** [opensearch-project/opensearch-net#388]
### Is your feature request related to a problem?
As a developer, I'm trying to reduce the amount of dependencies I'm taking on in my projects, and choose to use Microsofts default JSON serializer, System.Text.Json.

### What solution would you like?
An implementation of the `IOpenSearchSerializer` that uses System.Text.Json, that provides equivalent behaviour to the existing Newtonsoft implementation. 

### What alternatives have you considered?
At the moment, the fallback is very isolated usage of the Newtonsoft package.

### Do you have any additional context?
N/A

**Comment by Xtansia** [comment]
I've expanded the scope of this issue to instead be migrating away from `Utf8Json` to `System.Text.Json` as the primary serialization library of the clients.

**Comment by Djcarrillo6** [comment]
Hi @Xtansia, I'd like to request assignment on this issue.

**Comment by Xtansia** [comment]
@Djcarrillo6 You're more than welcome to have a go at this issue

**Comment by goldenc** [comment]
any update on this? we would like to move away from Newtonsoft.Json but its a bit tricky without a proper System.Text.Json open search implementation

**Comment by Xtansia** [comment]
> any update on this? we would like to move away from Newtonsoft.Json but its a bit tricky without a proper System.Text.Json open search implementation

@goldenc There's been no updates yet to the best of my knowledge, I'm unsure if @Djcarrillo6 made any progress. So would be great if anyone was willing to make contributions towards this goal.

**Comment by Djcarrillo6** [comment]
Yeah @Xtansia I've been running into road block after road block. If anyone else would like to also work on this, please feel free.

**Comment by Hailong-am** [comment]
## Utf8Json → System.Text.Json Migration Plan

**Goal:** Replace the vendored, unmaintained Utf8Json library with Microsoft's System.Text.Json (STJ), with a clear path to full AOT / Native AOT support.

---

### Phase 0 — STJ Opt-In (2.x)

Introduce STJ as a fully supported alternative without changing default behavior.

- Utf8Json remains the default serializer — zero breaking change
- `OSC_USE_STJ=true` environment variable switches to STJ
- `settings.UseSystemTextJson()` programmatic API does the same
- CI runs the full test suite under both engines to catch regressions on either path

**User impact:** Existing code is unaffected. Users can opt in to STJ, validate behavior, and report issues before the 3.0 default flip.

---

### Phase 1 — Decouple from Utf8Json Internals (2.x)

Prepare the codebase for eventual Utf8Json deletion, with no user-visible behavior change.

- Introduce a self-owned `[OpenSearchContractAttribute]` that replaces the Utf8Json-internal `[InterfaceDataContractAttribute]`. The resolver recognizes both so existing annotated types (428) continue to work unchanged. New types use the new attribute; existing types can be migrated gradually.
- Replace `Type.GetInterfaceMap()` with name-based interface property lookup — the only API that throws `NotSupportedException` on Native AOT
- Replace `MethodInfo.Invoke()` hot-path calls with cached compiled `Expression` delegates — AOT-compatible, ~10x faster
- Add `[RequiresUnreferencedCode]` at the STJ serializer entry point pointing to the Source Generator path
- Add `[DynamicallyAccessedMembers]` annotations so `dotnet publish --self-contained` works without crashes

**User impact:** `dotnet publish --self-contained` works on the STJ path. One trim warning is expected and documented.

---

### Phase 2 — Native AOT via Source Generator (2.x)

Eliminate all runtime reflection from the STJ path.

Replace the runtime `DataContractResolver` with a Roslyn `IIncrementalGenerator` that reads every `[OpenSearchContractAttribute]` type at build time, walks the interface hierarchy via Roslyn symbols, and emits static `JsonTypeInfo<T>` initialization code. Zero reflection at runtime.

**User impact:** `PublishAot=true` works. `[RequiresUnreferencedCode]` removed.

---

### Phase 3 — STJ Default + Delete Utf8Json (3.0)

Once the STJ path is mature across Phases 0–2:

- Flip default to STJ
- `OSC_USE_UTF8JSON=true` / `settings.UseUtf8Json()` as a short-lived escape hatch, both `[Obsolete]`
- Delete `src/OpenSearch.Net/Utf8Json/` (~20,000 lines)
- Remove all Utf8Json APIs and the dual-engine CI matrix

**User impact:** Smaller package, no abandoned dependency. Users on the escape hatch get a compile warning with upgrade guidance.

---

### Timeline

| Phase | Target | Deliverable |
|-------|--------|-------------|
| 0 | 2.x | STJ opt-in, dual-engine CI |
| 1 | 2.x | Self-owned contract attribute, trim-safe, AOT groundwork |
| 2 | 2.x | Source Generator, Native AOT |
| 3 | 3.0 | STJ default, Utf8Json deleted |

**Comment by xluo-aws** [comment]
@SamuelCox , @Xtansia , could you comment?

**Comment by SamuelCox** [comment]
Hi @xluo-aws , apologies what do you want us to comment on?

**Comment by xluo-aws** [comment]
The rollout plan @Hailong-am described above https://github.com/opensearch-project/opensearch-net/issues/388#issuecomment-5055962921

**Comment by Scrambles56** [comment]
Thank you all for the progress so far on this issue! I'll be adopting the STJ implementations into our product as they come along.

**Comment by xluo-aws** [comment]
> Thank you all for the progress so far on this issue! I'll be adopting the STJ implementations into our product as they come along.

Nice, let us know if you see any issues
